lib/segment/src/segment/partial_snapshot.rs

```diff
@@ -22,32 +22,62 @@ impl PartialSnapshotEntry for Segment {
         &self,
         temp_path: &Path,
         tar: &tar_ext::BuilderExt,
-        manifest: &SegmentManifest,
+        manifests: &SegmentManifests,
+        snapshotted_segments: &mut HashSet<String>,
     ) -> OperationResult<()> {
-        let segment_id = self
-            .current_path
-            .file_stem()
-            .and_then(|f| f.to_str())
-            .unwrap();
+        let segment_id = self.segment_id()?;
 
-        let updated_manifest = self.get_segment_manifest()?;
-        let updated_files = updated_files(manifest, &updated_manifest);
+        if !snapshotted_segments.insert(segment_id.into()) {
+            // Already snapshotted
+            return Ok(());
+        }
 
-        let tar = tar.descend(Path::new(&segment_id))?;
+        let updated_manifest = self.get_segment_manifest()?;
 
         let updated_manifest_json = serde_json::to_vec(&updated_manifest).map_err(|err| {
             OperationError::service_error(format!(
                 "failed to serialize segment manifest into JSON: {err}"
             ))
         })?;
 
+        let tar = tar.descend(Path::new(&segment_id))?;
         tar.blocking_append_data(&updated_manifest_json, Path::new("segment_manifest.json"))?;
 
-        snapshot_files(self, temp_path, &tar, |path| updated_files.contains(path))?;
+        match manifests.get(segment_id) {
+            Some(manifest) => {
+                let updated_files = updated_files(manifest, &updated_manifest);
+                snapshot_files(self, temp_path, &tar, |path| updated_files.contains(path))?;
+            }
+
+            None => {
+                snapshot_files(self, temp_path, &tar, |_| true)?;
+            }
+        }
+
         Ok(())
     }
 
+    fn collect_segment_manifests(&self, manifests: &mut SegmentManifests) -> OperationResult<()> {
+        manifests.add(self.get_segment_manifest()?);
+        Ok(())
+    }
+}
+
+impl Segment {
+    fn segment_id(&self) -> OperationResult<&str> {
+        self.current_path
+            .file_stem()
+            .and_then(|segment_dir| segment_dir.to_str())
```

Comment:
Shall we assert here it's a UUID, just in case we're doing some funny stuff we don't expect?

If so we might rename this to `segment_uuid()` to make it a bit more explicit.

Reply:
I'll add a debug insertion then, just in case!